File: libs/power-log-parser/src/lib/parsers/battlegrounds-trinket-selection-parser.ts
import { CardIds, CardType } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Action, FullEntity, Node } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class BattlegroundsTrinketSelectionParser implements ActionParser {
	readonly ParserName = 'BattlegroundsTrinketSelectionParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			this.StateFacade.IsBattlegrounds() &&
			node.Type === FullEntity &&
			(node.Object as FullEntity).GetCardType() === (CardType.BATTLEGROUND_TRINKET as number)
		);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		if (node.Parent == null || node.Parent.Type !== Action) {
			return null;
		}

		const parentAction = node.Parent.Object as Action;
		const entity = this.GameState.CurrentEntities.get(parentAction.Entity);
		if (
			entity?.CardId !== CardIds.LesserTrinketToken_BG30_Trinket_1st &&
			entity?.CardId !== CardIds.GreaterTrinket_BG30_Trinket_2nd
		) {
			// Don't return null here - we also want to show data for trinkets created by other trinkets or hero powers
		}

		const controllerId = entity!.GetEffectiveController();
		const options = parentAction.Data.filter((d) => d instanceof FullEntity)
			.map((d) => d as unknown as FullEntity)
			.filter((f) => f.GetCardType() === (CardType.BATTLEGROUND_TRINKET as number))
			.map((f) => ({
				CardId: f.CardId,
				Cost: f.GetCost(),
			}));
		if (options.length !== 4) {
			return null;
		}

		return [
			GameEventProvider.Create(
				parentAction.TimeStamp,
				'BATTLEGROUNDS_TRINKET_SELECTION',
				GameEventHelper.CreateProvider(
					'BATTLEGROUNDS_TRINKET_SELECTION',
					entity!.CardId,
					controllerId,
					entity!.Entity,
					this.StateFacade,
					{
						Options: options,
					},
				),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/BattlegroundsTrinketSelectionParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Linq;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using static HearthstoneReplays.Events.CardIds;

namespace HearthstoneReplays.Events.Parsers
{
    public class BattlegroundsTrinketSelectionParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public BattlegroundsTrinketSelectionParser(ParserState ParserState, StateFacade stateFacade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = stateFacade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            // Because the BLOCK is corrupted because of the choices, so we use another technique
            return stateType == StateType.PowerTaskList
                && StateFacade.IsBattlegrounds()
                && node.Type == typeof(FullEntity)
                && (node.Object as FullEntity).GetCardType() == (int)CardType.BATTLEGROUND_TRINKET;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var fullEntity = node.Object as FullEntity;
            if (node.Parent == null || node.Parent.Type != typeof(Action))
            {
                return null;
            }

            var parentAction = node.Parent.Object as Action;
            var entity = GameState.CurrentEntities.GetValueOrDefault(parentAction.Entity);
            if (entity?.CardId != CardIds.LesserTrinketToken_BG30_Trinket_1st && entity?.CardId != CardIds.GreaterTrinket_BG30_Trinket_2nd)
            {
                // Don't do this, as we also want to show the data for trinkets created by other trinkets or hero powers
                //return null;
            }

            var controllerId = entity.GetEffectiveController();
            var options = parentAction.Data
                .Where(d => d is FullEntity)
                .Select(d => d as FullEntity)
                .Where(f => f.GetCardType() == (int)CardType.BATTLEGROUND_TRINKET)
                .Select(f =>
                {
                    return new
                    {
                        CardId = f.CardId,
                        Cost = f.GetCost(),
                    };
                })
                .ToList();
            if (options.Count != 4)
            {
                return null;
            }

            return new List<GameEventProvider> { GameEventProvider.Create(
                parentAction.TimeStamp,
                "BATTLEGROUNDS_TRINKET_SELECTION",
                GameEvent.CreateProvider(
                "BATTLEGROUNDS_TRINKET_SELECTION",
                    entity.CardId,
                    controllerId,
                    entity.Entity,
                    StateFacade,
                    //null,
                    new {
                        Options = options
                    }),
                true,
                node) };
        }
    }
}
